### PR TITLE
IBP-3088-ImportExportEnvironments

### PR DESCRIPTION
**Pull request by mderamos**

### CODE REVIEW

File: src/main/java/org/generationcp/middleware/service/impl/dataset/DatasetServiceImpl.java [hunk 1]
```diff
@@ -158,6 +168,9 @@ public class DatasetServiceImpl implements DatasetService {
 		TermId.EXPERIMENT_DESIGN_FACTOR.getId());
 	private static final String SUM_OF_SAMPLES = "SUM_OF_SAMPLES";
 
+	private static final List<Integer> EXP_DESIGN_VARIABLES =
```
Comment (abatac) [inline src/main/java/org/generationcp/middleware/service/impl/dataset/DatasetServiceImpl.java:171]: I think you can just use the existing final AppConstants.EXP_DESIGN_VARIABLES:

` List<Integer> expDesignConstants = AppConstants.EXP_DESIGN_VARIABLES.getIntegerList()`

File: src/main/java/org/generationcp/middleware/service/impl/dataset/DatasetServiceImpl.java [hunk 2]
```diff
@@ -1254,6 +1270,112 @@ public Map<String, ObservationUnitRow> getObservationUnitsAsMap(
...
+			}
+
+			final Map<String, Geolocation> geolocationMap = this.daoFactory.getGeolocationDao().findInDataSet(datasetId).stream()
+					.collect(Collectors.toMap(geolocation -> geolocation.getDescription(), Function.identity()));
+
+			final Map<String, ObservationUnitRow> currentData =
+					this.daoFactory.getExperimentDao().getObservationUnitsAsMap(datasetId, measurementVariableList,
+							observationUnitIds);
+
+			final List<GeolocationProperty> geolocationProperties = this.daoFactory.getGeolocationPropertyDao()
```
Comment (abatac) [inline src/main/java/org/generationcp/middleware/service/impl/dataset/DatasetServiceImpl.java:1308]: Extract this to a private method e.g. getGeolocationProperties(instanceIds);

File: src/main/java/org/generationcp/middleware/service/impl/dataset/DatasetServiceImpl.java [hunk 3]
```diff
@@ -1254,6 +1270,112 @@ public Map<String, ObservationUnitRow> getObservationUnitsAsMap(
...
 	}
 
+	@Override
+	public void importEnvironmentVariableValues(final Integer studyId, final Integer datasetId, final Table<String, String, String> table) {
+		final List<MeasurementVariable> measurementVariableList =
+				this.getDatasetMeasurementVariablesByVariableType(datasetId,
+						Arrays.asList(VariableType.ENVIRONMENT_DETAIL.getId(), VariableType.ENVIRONMENT_CONDITION.getId()));
+
+		if (!measurementVariableList.isEmpty()) {
+			final Map<String, MeasurementVariable> mappedVariables = new HashMap<>();
```
Comment (abatac) [inline src/main/java/org/generationcp/middleware/service/impl/dataset/DatasetServiceImpl.java:1280]: Extract this to a private method e.g. createMappedVariables(measurementVariableList)

File: src/main/java/org/generationcp/middleware/service/impl/dataset/DatasetServiceImpl.java [hunk 4]
```diff
@@ -1254,6 +1270,112 @@ public Map<String, ObservationUnitRow> getObservationUnitsAsMap(
...
+			final Map<String, Map<Integer, GeolocationProperty>> instanceNoPropertyMap = new HashMap<>();
+			for(final GeolocationProperty property: geolocationProperties) {
+				final String instanceNumber = property.getGeolocation().getDescription();
+				instanceNoPropertyMap.putIfAbsent(instanceNumber, new HashMap<>());
+				instanceNoPropertyMap.get(instanceNumber).put(property.getTypeId(), property);
+			}
+
+
+			final List<String> locationNames = new ArrayList<>();
+			Map<String, Integer> locationNameLocationIdMap = new HashMap<>();
```
Comment (abatac) [inline src/main/java/org/generationcp/middleware/service/impl/dataset/DatasetServiceImpl.java:1319]: Extract this to a private method e.g.  getLocationNameLocationIdMap(table)

**Review by abatac — CHANGES_REQUESTED** [review]

File: src/main/java/org/generationcp/middleware/service/impl/dataset/DatasetServiceImpl.java (hunk 1 above)
Comment (mderamos) [reply src/main/java/org/generationcp/middleware/service/impl/dataset/DatasetServiceImpl.java:171]: AppConstants are in Commons which can't be accessed in MW

**Review by abatac — APPROVED** [review]

**Pull request merged**